File: output/cpp/tests/18_transition_exit_args.cpp
#include <string>
#include <unordered_map>
#include <vector>
#include <any>
#include <memory>
#include <functional>


#include <iostream>
#include <string>
#include <cassert>
#include <vector>
#include <algorithm>

class TransitionExitArgsFrameEvent {
public:
    std::string _message;
    std::unordered_map<std::string, std::any> _parameters;

    TransitionExitArgsFrameEvent(const std::string& message, std::unordered_map<std::string, std::any> params = {})
        : _message(message), _parameters(std::move(params)) {}
};

class TransitionExitArgsFrameContext {
public:
    TransitionExitArgsFrameEvent _event;
    std::any _return;
    std::unordered_map<std::string, std::any> _data;

    TransitionExitArgsFrameContext(TransitionExitArgsFrameEvent event, std::any default_return = {})
        : _event(std::move(event)), _return(std::move(default_return)) {}
};

class TransitionExitArgsCompartment {
public:
    std::string state;
    std::unordered_map<std::string, std::any> state_args;
    std::unordered_map<std::string, std::any> state_vars;
    std::unordered_map<std::string, std::any> enter_args;
    std::unordered_map<std::string, std::any> exit_args;
    std::unique_ptr<TransitionExitArgsFrameEvent> forward_event;
    std::unique_ptr<TransitionExitArgsCompartment> parent_compartment;

    explicit TransitionExitArgsCompartment(const std::string& state) : state(state) {}

    std::unique_ptr<TransitionExitArgsCompartment> clone() const {
        auto c = std::make_unique<TransitionExitArgsCompartment>(state);
        c->state_args = state_args;
        c->state_vars = state_vars;
        c->enter_args = enter_args;
        c->exit_args = exit_args;
        return c;
    }
};

class TransitionExitArgs {
private:
    std::vector<std::unique_ptr<TransitionExitArgsCompartment>> _state_stack;
    std::unique_ptr<TransitionExitArgsCompartment> __compartment;
    std::unique_ptr<TransitionExitArgsCompartment> __next_compartment;
    std::vector<TransitionExitArgsFrameContext> _context_stack;

    void __kernel(TransitionExitArgsFrameEvent& __e) {
        __router(__e);
        while (__next_compartment) {
            auto next_compartment = std::move(__next_compartment);
            TransitionExitArgsFrameEvent exit_event("<$");
            __router(exit_event);
            __compartment = std::move(next_compartment);
            if (!__compartment->forward_event) {
                TransitionExitArgsFrameEvent enter_event("$>");
                __router(enter_event);
            } else {
                auto forward_event = std::move(__compartment->forward_event);
                if (forward_event->_message == "$>") {
                    __router(*forward_event);
                } else {
                    TransitionExitArgsFrameEvent enter_event("$>");
                    __router(enter_event);
                    __router(*forward_event);
                }
            }
        }
    }

    void __router(TransitionExitArgsFrameEvent& __e) {
        const auto& state_name = __compartment->state;
        if (state_name == "Active") {
            _state_Active(__e);
        } else if (state_name == "Done") {
            _state_Done(__e);
        }
    }

    void __transition(std::unique_ptr<TransitionExitArgsCompartment> next) {
        __next_compartment = std::move(next);
    }

    void _state_Done(TransitionExitArgsFrameEvent& __e) {
        if (__e._message == "$>") {
            event_log.push_back("enter:done");
        } else if (__e._message == "get_log") {
            _context_stack.back()._return = std::any(event_log);
            return;;
        }
    }

    void _state_Active(TransitionExitArgsFrameEvent& __e) {
        if (__e._message == "<$") {
            auto reason = std::any_cast<std::string>(__compartment->exit_args["0"]);
            auto code = std::any_cast<int>(__compartment->exit_args["1"]);
            event_log.push_back(std::string("exit:") + reason + ":" + std::to_string(code));
        } else if (__e._message == "get_log") {
            _context_stack.back()._return = std::any(event_log);
            return;;
        } else if (__e._message == "leave") {
            event_log.push_back("leaving");
            __compartment->exit_args["0"] = std::any(std::string("cleanup"));
            __compartment->exit_args["1"] = std::any(42);
            auto __new_compartment = std::make_unique<TransitionExitArgsCompartment>("Done");
            __new_compartment->parent_compartment = __compartment->clone();
            __transition(std::move(__new_compartment));
            return;
        }
    }

public:
    std::vector<std::string> event_log = {};

    TransitionExitArgs() {
        __compartment = std::make_unique<TransitionExitArgsCompartment>("Active");
        TransitionExitArgsFrameEvent __frame_event("$>");
        TransitionExitArgsFrameContext __ctx(std::move(__frame_event));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void leave() {
        TransitionExitArgsFrameEvent __e("leave");
        TransitionExitArgsFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    std::vector<std::string> get_log() {
        TransitionExitArgsFrameEvent __e("get_log");
        TransitionExitArgsFrameContext __ctx(std::move(__e), std::any(std::vector<std::string>()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<std::vector<std::string>>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }
};

int main() {
    printf("=== Test 18: Transition Exit Args ===\n");
    TransitionExitArgs s;

    auto log = s.get_log();
    if (log.size() != 0) {
        printf("FAIL: Expected empty log\n");
        assert(false);
    }

    s.leave();
    log = s.get_log();
    if (std::find(log.begin(), log.end(), "leaving") == log.end()) {
        printf("FAIL: Expected 'leaving' in log\n");
        assert(false);
    }
    if (std::find(log.begin(), log.end(), "exit:cleanup:42") == log.end()) {
        printf("FAIL: Expected 'exit:cleanup:42' in log\n");
        assert(false);
    }
    if (std::find(log.begin(), log.end(), "enter:done") == log.end()) {
        printf("FAIL: Expected 'enter:done' in log\n");
        assert(false);
    }
    printf("Log after transition: leaving, exit:cleanup:42, enter:done\n");

    printf("PASS: Transition exit args work correctly\n");
    return 0;
}

File: output/go/tests/18_transition_exit_args.go

package main

import "fmt"
import "os"

type TransitionExitArgsFrameEvent struct {
    _message    string
    _parameters map[string]any
}

type TransitionExitArgsFrameContext struct {
    _event  TransitionExitArgsFrameEvent
    _return any
    _data   map[string]any
}

type TransitionExitArgsCompartment struct {
    state            string
    stateArgs        map[string]any
    stateVars        map[string]any
    enterArgs        map[string]any
    exitArgs         map[string]any
    forwardEvent     *TransitionExitArgsFrameEvent
    parentCompartment *TransitionExitArgsCompartment
}

func newTransitionExitArgsCompartment(state string) *TransitionExitArgsCompartment {
    return &TransitionExitArgsCompartment{
        state:    state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
    }
}

func (c *TransitionExitArgsCompartment) copy() *TransitionExitArgsCompartment {
    nc := &TransitionExitArgsCompartment{
        state: c.state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
        forwardEvent:     c.forwardEvent,
        parentCompartment: c.parentCompartment,
    }
    for k, v := range c.stateArgs { nc.stateArgs[k] = v }
    for k, v := range c.stateVars { nc.stateVars[k] = v }
    for k, v := range c.enterArgs { nc.enterArgs[k] = v }
    for k, v := range c.exitArgs { nc.exitArgs[k] = v }
    return nc
}

type TransitionExitArgs struct {
    _state_stack []*TransitionExitArgsCompartment
    __compartment *TransitionExitArgsCompartment
    __next_compartment *TransitionExitArgsCompartment
    _context_stack []TransitionExitArgsFrameContext
    log []string
}

func NewTransitionExitArgs() *TransitionExitArgs {
    s := &TransitionExitArgs{}
    s._state_stack = make([]*TransitionExitArgsCompartment, 0)
    s._context_stack = make([]TransitionExitArgsFrameContext, 0)
    s.__compartment = newTransitionExitArgsCompartment("Active")
    s.__next_compartment = nil
    __frame_event := TransitionExitArgsFrameEvent{_message: "$>", _parameters: nil}
    __ctx := TransitionExitArgsFrameContext{_event: __frame_event, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return s
}

func (s *TransitionExitArgs) __kernel(__e *TransitionExitArgsFrameEvent) {
    s.__router(__e)
    for s.__next_compartment != nil {
        next_compartment := s.__next_compartment
        s.__next_compartment = nil
        exit_event := &TransitionExitArgsFrameEvent{_message: "<$", _parameters: s.__compartment.exitArgs}
        s.__router(exit_event)
        s.__compartment = next_compartment
        if s.__compartment.forwardEvent == nil {
            enter_event := &TransitionExitArgsFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
            s.__router(enter_event)
        } else {
            forward_event := s.__compartment.forwardEvent
            s.__compartment.forwardEvent = nil
            if forward_event._message == "$>" {
                s.__router(forward_event)
            } else {
                enter_event := &TransitionExitArgsFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
                s.__router(enter_event)
                s.__router(forward_event)
            }
        }
    }
}

func (s *TransitionExitArgs) __router(__e *TransitionExitArgsFrameEvent) {
    switch s.__compartment.state {
    case "Active":
        s._state_Active(__e)
    case "Done":
        s._state_Done(__e)
    }
}

func (s *TransitionExitArgs) __transition(next *TransitionExitArgsCompartment) {
    s.__next_compartment = next
}

func (s *TransitionExitArgs) Leave() {
    __e := TransitionExitArgsFrameEvent{_message: "Leave"}
    __ctx := TransitionExitArgsFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *TransitionExitArgs) GetLog() []string {
    __e := TransitionExitArgsFrameEvent{_message: "GetLog"}
    __ctx := TransitionExitArgsFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result []string
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.([]string) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *TransitionExitArgs) _state_Done(__e *TransitionExitArgsFrameEvent) {
    if __e._message == "$>" {
        s.log = append(s.log, "enter:done")
    } else if __e._message == "GetLog" {
        s._context_stack[len(s._context_stack)-1]._return = s.log
        return
    }
}

func (s *TransitionExitArgs) _state_Active(__e *TransitionExitArgsFrameEvent) {
    if __e._message == "<$" {
        reason := s.__compartment.exitArgs["0"].(string)
        _ = reason
        code := s.__compartment.exitArgs["1"].(int)
        _ = code
        s.log = append(s.log, fmt.Sprintf("exit:%s:%d", reason, code))
    } else if __e._message == "GetLog" {
        s._context_stack[len(s._context_stack)-1]._return = s.log
        return
    } else if __e._message == "Leave" {
        s.log = append(s.log, "leaving")
        s.__compartment.exitArgs["0"] = "cleanup"
        s.__compartment.exitArgs["1"] = 42
        __compartment := newTransitionExitArgsCompartment("Done")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    }
}

func contains(slice []string, item string) bool {
	for _, v := range slice {
		if v == item {
			return true
		}
	}
	return false
}

func main() {
	fmt.Println("=== Test 18: Transition Exit Args ===")
	sm := NewTransitionExitArgs()

	// Initial state is Active
	log := sm.GetLog()
	if len(log) != 0 {
		fmt.Printf("FAIL: Expected empty log, got %v\n", log)
		os.Exit(1)
	}

	// Leave - should call exit handler with args
	sm.Leave()
	log = sm.GetLog()
	if !contains(log, "leaving") {
		fmt.Printf("FAIL: Expected 'leaving' in log, got %v\n", log)
		os.Exit(1)
	}
	if !contains(log, "exit:cleanup:42") {
		fmt.Printf("FAIL: Expected 'exit:cleanup:42' in log, got %v\n", log)
		os.Exit(1)
	}
	if !contains(log, "enter:done") {
		fmt.Printf("FAIL: Expected 'enter:done' in log, got %v\n", log)
		os.Exit(1)
	}
	fmt.Printf("Log after transition: %v\n", log)

	fmt.Println("PASS: Transition exit args work correctly")
}
